spec-tests/src/test/java/org/candlepin/spec/environments/EnvironmentCertV3SpecTest.java

```diff
@@ -0,0 +1,280 @@
+        adminClient.owners().createPool(ownerKey, Pools.random(product));
+
+        JsonNode ents = consumerClient.consumers().bindProduct(consumer.getUuid(), product.getId());
+
+        assertThat(ents).singleElement();
+        String cert = getCertFromEnt(ents.get(0));
+        assertThat(CertificateUtil.getDerValueFromExtension(cert, ENT_VERSION_OID).toString())
+            .isEqualTo("3.4");
+        JsonNode entCert = CertificateUtil.decodeAndUncompressCertificate(cert, ApiClient.MAPPER);
+        assertThat(entCert.get("products")).singleElement();
+        assertThat(entCert.get("products").get(0).get("content"))
+            .singleElement()
+            .returns(promotedContent.getId(), cont -> cont.get("id").asText());
+    }
+
+    @Test
+    public void shouldRegenerateCertificatesAfterPromotingAndDemotingContent() {
+        ApiClient adminClient = ApiClients.admin();
+        OwnerDTO owner = adminClient.owners().createOwner(Owners.random());
+        String ownerKey = owner.getKey();
+        ApiClient orgAdmin = ApiClients.basic(UserUtil.createAdminUser(adminClient, owner));
+        EnvironmentDTO env = orgAdmin.owners().createEnv(ownerKey, Environments.random());
+        ConsumerDTO consumer = orgAdmin.consumers().createConsumer(Consumers.random(owner)
+            .addEnvironmentsItem(env)
+            .facts(Map.of("system.certificate_version", "3.1")));
+        assertThat(consumer)
+            .isNotNull()
+            .extracting(ConsumerDTO::getEnvironments, as(collection(EnvironmentDTO.class)))
+            .singleElement();
+
+        ApiClient consumerClient = ApiClients.ssl(consumer);
+        ProductDTO product = adminClient.ownerProducts().createProductByOwner(ownerKey, Products.randomEng());
+        ContentDTO promotedContent = adminClient.ownerContent().createContent(ownerKey, Contents.random());
+        ContentDTO notPromotedContent = adminClient.ownerContent().createContent(ownerKey, Contents.random());
+        adminClient.ownerProducts().addContent(ownerKey, product.getId(), promotedContent.getId(), true);
+        adminClient.ownerProducts().addContent(ownerKey, product.getId(), notPromotedContent.getId(), true);
+        promoteContentToEnvironment(orgAdmin, env.getId(), promotedContent, true);
+        adminClient.owners().createPool(ownerKey, Pools.random(product));
+
+        JsonNode ents = consumerClient.consumers().bindProduct(consumer.getUuid(), product.getId());
+
+        assertThat(ents).singleElement();
+        String cert = getCertFromEnt(ents.get(0));
+        assertThat(CertificateUtil.getDerValueFromExtension(cert, ENT_VERSION_OID).toString())
+            .isEqualTo("3.4");
+        JsonNode entCert = CertificateUtil.decodeAndUncompressCertificate(cert, ApiClient.MAPPER);
+        JsonNode products = entCert.get("products");
+        assertThat(products).singleElement();
+        assertThat(products.get(0).get("content"))
```

Comment:
This could be chained off the dangling `.singleElement()` to avoid having to restart the chain with `.get(0)`:

```
assertThat(products)
  .singleElement()
  .extracting(pnode -> pnode.get("content"))
  .singleElement()
  .returns(...);
```

This is applicable in multiple locations.